### PR TITLE
[Refactoring] Adding a sign/ package with Schnorr

### PR DESCRIPTION
**Pull request by nikkolasg**
This PR paves the way for some functionality needed in VSS later. It adds a `sign` package with a `schnorr` signature framework, the one from `dedis/onet`.

### CODE REVIEW

File: sign/schnorr.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	if _, err := e.MarshalTo(&b); err != nil {
+		return nil, err
+	}
+	if _, err := s.MarshalTo(&b); err != nil {
+		return nil, err
+	}
+	return b.Bytes(), nil
+}
+
+// VerifySchnorr verifies a given Schnorr signature. It returns nil iff the
```
Comment (ineiti) [inline sign/schnorr.go]: s/iff/if/

File: sign/schnorr.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+// given signature is valid.  NOTE: this signature scheme is malleable because
+// the response's unmarshalling is done directly into a big.Int modulo (see
+// nist.Int).
+func VerifySchnorr(suite abstract.Suite, public abstract.Point, msg, sig []byte) error {
+	var challenge = suite.Scalar()
+	var response = suite.Scalar()
+	scalarSize := challenge.MarshalSize()
+	sigSize := scalarSize * 2
```
Comment (ineiti) [inline sign/schnorr.go]: style: why use twice`var` and twice `:=`? I propose to use `:=` in functions and `var` in package-global-variables.

File: sign/schnorr.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/random"
+)
+
+// HashFunc is the hash function used to hash the message for the signature
+// generation and verification
+var HashFunc = sha256.New
```
Comment (ineiti) [inline sign/schnorr.go]: style: global variable that is used only once - can go into the local `hash`-function, IMO.

File: sign/schnorr_test.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package sign
+
+import (
+	"testing"
+
+	"github.com/dedis/crypto/config"
+	"github.com/dedis/crypto/ed25519"
+)
+
+func TestSchnorrSignature(t *testing.T) {
```
Comment (ineiti) [inline sign/schnorr_test.go]: does not test failing parts of the schnorr-verification.

File: sign/schnorr.go (hunk 1 above)
Comment (Daeinar) [inline sign/schnorr.go]: In math "iff" sometimes means "if and only if", see https://en.wikipedia.org/wiki/If_and_only_if

Not sure if it is the case here or if it's just a typo.

File: sign/schnorr.go (hunk 3 above)
Comment (nikkolasg) [inline sign/schnorr.go]: In fact, I should use the `suite.Hash()` function, don't you think ?

File: sign/schnorr.go (hunk 1 above)
Comment (nikkolasg) [inline sign/schnorr.go]: Exactly, that's what I meant with this `iff`: it returns nil if and only if the signature is valid.

File: sign/schnorr_test.go (hunk 4 above)
Comment (nikkolasg) [inline sign/schnorr_test.go]: Yep, will do.

**Comment by nikkolasg** [comment]
> is it better to have a schnorr-structure (like I ask in the challenge), or to have only package-functions?

I have absolutely no arguments in favor of having a struct instead of a slice of bytes but a lot against :D 
 - one should not mess up with a signature anyway so no need to expose its fields
 - it's a new struct that user must learn about, to integrate with their code vs a plain simple slice of byte (  "complexification" of the library )
 - the usage of a struct makes the encoding responsible to the user. that means it can potentially be non valid against EdDSA signature verification and CoSi verification. 
 - we used slice of bytes for CoSi and EdDSA

File: sign/schnorr.go (hunk 1 above)
Comment (ineiti) [inline sign/schnorr.go]: OK, iff you're all doing mathy stuff, then so be it...

File: sign/schnorr_test.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	// wrong response
+	wrResp := make([]byte, len(s))
+	copy(wrResp[:32], random.Bytes(32, random.Stream))
+	copy(wrResp[32:], s[32:])
+	assert.Error(t, VerifySchnorr(suite, kp.Public, msg, wrResp))
+
+	// wrong public key
+	wrKp := config.NewKeyPair(suite)
+	assert.Error(t, VerifySchnorr(suite, wrKp.Public, msg, s))
+
```
Comment (ineiti) [inline sign/schnorr_test.go]: style: delete empty line.

**Pull request merged**